docs/internal/types.md

```diff
@@ -486,14 +486,29 @@ interface Addable
 {
   +(Addable & readonly, Addable & readonly) : Addable & imm;
 }
-class Foo : Addable { ... }
-class Bar : Addable { ... }
-class Baz : Addable { ... }
+class Foo : Addable
+{
+  // Implements interface and operates on Foo only (as expected)
+  +(Foo & readonly, Foo & readonly) : Foo & imm { ... }
+}
+class Bar : Addable
+{
+  // Implements interface and operates on Bar only (as expected)
+  +(Bar & readonly, Bar & readonly) : Bar & imm { ... }
+}
+class Baz : Addable
+{
+  // Implements interface but operates on different types
+  // This is perfectly valid for the interface but it isn't what we want
+  +(Foo & readonly, Bar & readonly) : Baz & imm { ... }
+}
 
 function() {
-  foo = Foo;
-  bar = Bar;
+  var foo = Foo;
+  var bar = Bar;
+  // Type error.  Foo's + implementation takes parameters of type `Foo` not arbitrary addable things.
+  var addable : Addable = foo;
   // This matches the interface, but not Foo's (or Bar's) implementation
-  baz : Baz & imm = Addable::+(foo, bar);
+  var baz : Baz & imm = foo + bar;
```

Comment:
This doesn't look like it works either.  You need to be calling `Baz::+`, but you're going to call `Foo::+`, which doesn't support these types.

Reply:
IIUC, this calls `Foo::+(foo, bar)` because the descriptor comes from the left-hand side, not the assignment type.